pkg/standalone-utils/contracts/test/MockTetuSmartVault.sol

```diff
@@ -0,0 +1,109 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity >=0.7.0 <0.9.0;
+
+import "@balancer-labs/v2-interfaces/contracts/standalone-utils/ITetuSmartVault.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/test/TestToken.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+
+import "./MockTetuStrategy.sol";
+import "../relayer/special/TetuShareValueHelper.sol";
+
+contract MockTetuSmartVault is ITetuSmartVault, TestToken, TetuShareValueHelper {
+    using SafeERC20 for IERC20;
+    using FixedPoint for uint256;
+
+    IERC20 public underlyingAsset;
+    uint256 private immutable _underlyingDecimals;
+    uint256 private _underlyingBalanceInVault = 0;
+    MockTetuStrategy private immutable _tetuStrategy;
+    uint256 _desiredRate;
+
+    constructor(
+        string memory name,
+        string memory symbol,
+        uint8 decimals,
+        address _underlyingAsset,
+        MockTetuStrategy tetuStrategy
+    ) TestToken(name, symbol, decimals) {
+        underlyingAsset = IERC20(_underlyingAsset);
+        _underlyingDecimals = decimals;
+        _tetuStrategy = tetuStrategy;
+    }
+
+    function getPricePerFullShare() external pure override returns (uint256) {
+        revert("Should not call this");
+    }
+
+    // Should pass rate with decimals from underlyingToken
+    function setRate(uint256 newRate) public {
+        // stores latest rate, so the balances are recalculated after deposit and withdraw
+        _desiredRate = newRate;
+        uint256 totalSupply = this.totalSupply();
+        // arbitrary number, just to make sure that both Vault and Invested values compose the rate.
+        uint8 vaultInvestedRatio = 3;
+        uint256 totalBalance = (newRate * totalSupply) / 10**_underlyingDecimals;
+        _underlyingBalanceInVault = totalBalance / vaultInvestedRatio;
+        _tetuStrategy.setInvestedUnderlyingBalance(totalBalance - _underlyingBalanceInVault);
+    }
+
+    function underlyingBalanceInVault() external view override returns (uint256) {
+        return _underlyingBalanceInVault;
+    }
+
+    function underlyingBalanceWithInvestmentForHolder(address) external view override returns (uint256) {
+        return underlyingAsset.balanceOf(address(this));
+    }
+
+    function deposit(uint256 amount) external override {
+        IERC20(underlyingAsset).safeTransferFrom(msg.sender, address(this), amount);
+        // Makes sure rate balances are correctly set before calculating wrappedAmount
+        setRate(_desiredRate);
```

Comment:
I understand why this is necessary after total supply changes (i.e., mint/burn), but why is it needed here?

Reply:
Because Tetu's token can be minted outside of the contract, so we need to reset the balances to keep the rate in the right value before calculating wrappedAmount.

Reply:
I see. In that case, can we override `_mint()` and `_burn()` instead? That way these updates occur automatically whenever they are needed. It could look like this, for example:

```
function _mint(address account, uint256 amount) internal virtual override {
    super._mint(account, amount);
    setRate(_desiredRate);
}
```